### DISCUSSION TITLE
custom Fields config implementation, cant get nested value type on dynamic name

### DISCUSSION THREAD

**Discussion by borm** [react-hook-form/react-hook-form#7682] [Q&A]
**Is your feature request related to a problem? Please describe.**
Hey there, im trying to do something like `fieldsConfig` for map my fields and render then by `switch` by type and return needed component, but i cant get value of nested property when my field name is dynamic

**Describe the solution you'd like**
the problem is probably somewhere here,  (solution from [stackoverflow](https://stackoverflow.com/questions/51502671/typescript-getting-the-type-of-object-key-inside-object-itself#answer-51503091))
```
type PossibleFields<T>
```
[codesandbox](https://codesandbox.io/s/react-hook-form-nestedvalue-forked-xio6s?file=/src/index.ts)

**update**
found temporary solution, look at **diff**

```diff
import {
  FieldPath,
  FieldValues,
  UseControllerProps,
  UseControllerReturn
} from "react-hook-form";

interface HookField<
  T extends FieldValues = FieldValues,
  K extends FieldPath<T> = FieldPath<T>> extends UseControllerProps<T, K> {
+ key: K;
  render?: (props: UseControllerReturn<T, K>) => void;
}

type PossibleFields<T> = FieldPath<T> extends infer K
  ? K extends FieldPath<T>
    ? HookField<T, K>
    : never
  : never;

export function hookFields<T extends FieldValues = FieldValues>(
  fields: Array<PossibleFields<T>>
) {
  // return fields.map(() => {
  // logic to render different fields by type
  // will be here
  // })
  return fields;
}

interface Values {
  test: string;
  arr: Array<{
    id: string;
  }>;
}

const consoleString = (value: string) => console.log(value);

hookFields<Values>([
  {
    name: "test",
    render: ({ field: { value } }) => {
      consoleString(value);
    }
  }
]);

[].map((k, index) => {
  hookFields<Values>([
    {
      name: `arr.0.id`,
      render: ({ field: { value } }) => {
        consoleString(value);
      }
    },
    {
      name: `arr.${index}.id` as const,
+     key: `arr.0.id`,
      render: ({ field: { value } }) => {
        // var value: any
        // Binding element 'value' implicitly has an 'any' type.ts(7031)
        consoleString(value);
      }
    }
  ]);
});

```

**Additional context**
with normal **field name**  as `string`, everything is fine and im getting field value as well

![image](https://user-images.githubusercontent.com/4027842/151412512-2efcf49d-da33-443f-830a-d8a614370c39.png)

but when **field name** is generated **dynamically** i cant get valid value type, is `any`

![image](https://user-images.githubusercontent.com/4027842/151412628-8c4e6949-06aa-4733-b613-4f6de8e667f6.png)

i know is not a `hook-form` issue, by i would like to get ur help because u know better ur library
sorry for ask here, but im lost all my day already and cant find any solution

**update**
just interesting case,  typescript understand what dynamic key is exist in array `arr.${number}.id`
``` typescript
type GetFieldKeys<T> = T extends Array<infer K>
  ? K extends { name: string }
    ? K['name']
    : never
  : never;

[].map((k, index) => {
  const fields = [
    {
      name: `arr.0.id`,
      render: ({ field: { value } }) => {
        consoleString(value);
      },
    },
    {
      name: `arr.${index}.id` as const,
      render: ({ field: { value } }) => {
        consoleString(value);
      },
    },
  ];
  hookFields<Values>(fields);
  type Keys = GetFieldKeys<typeof fields>;
});
```

![image](https://user-images.githubusercontent.com/4027842/151434038-e72d109a-c74f-4782-a3f8-c0b1d33307c3.png)

**Comment by JamesWong1999** [comment]
This should be a question instead of feature request right?

**Reply by borm** [reply]
yes, thanks for moving that

**Reply by bluebill1049** [reply]
hey, @borm I think this is rather difficult or probably not possible with V7, we are working on v8 at the moment: https://github.com/react-hook-form/react-hook-form/discussions/7354 with the improvement should be able to resolve this type of difficulty.

**Reply by borm** [reply]
hey @bluebill1049 , i guess its not a V7 issue, just something wrong on typescript part, or do u think its somewhere in `FieldPath` interface?

**Reply by borm** [reply]
ah, just saw ur linked discussion, so yes, i see, u have plan to do new api

**Reply by bluebill1049** [reply]
yes @felixschorer is driving this right now and made some really impressive progress.

**Reply by felixschorer** [reply]
@borm please have a look at #7389, specifically the "Other Use Cases" section. That seems to be exactly what you're trying to do.

**Comment by bluebill1049** [comment]
I will have a look at it after work.
